---
title: "talk04 练习与作业"
documentclass: ctexart
output:
  rticles::ctex:
    fig_caption: yes
    number_sections: yes
    toc: true
    toc_depth: 2
  word_document: default
  html_document:
    df_print: paged
knit: (
  function(inputFile, encoding) { 

    pSubTitle <- 'talk04-homework'

    rmarkdown::render( 
      input       = inputFile, 
      encoding    = encoding, 
      params      = list(sub_title = pSubTitle),      
      output_file = pSubTitle) })
---

## 练习和作业说明

将相关代码填写入以 \`\`\`{r} \`\`\` 标志的代码框中，运行并看到正确的结果；

完成后，用工具栏里的"Knit"按键生成PDF文档；

**将PDF文档**改为：`姓名-学号-talk04作业.pdf`，并提交到老师指定的平台/钉群。

## Talk04 内容回顾

待写 ...

## 练习与作业：用户验证

请运行以下命令，验证你的用户名。

**如你当前用户名不能体现你的真实姓名，请改为拼音后再运行本作业！**

```{r}
Sys.info()[["user"]]
Sys.getenv("HOME")
```

## 练习与作业1：R session管理

------------------------------------------------------------------------

### **完成以下操作**

-   定义一些变量（比如 x, y , z 并赋值；内容随意）
-   从外部文件装入一些数据（可自行创建一个4行5列的数据，内容随意）
-   保存workspace 到.RData
-   列出当前工作空间内的所有变量
-   删除当前工作空间内所有变量
-   从.RData文件恢复保存的数据
-   再次列出当前工作空间内的所有变量，以确认变量已恢复
-   随机删除两个变量
-   再次列出当前工作空间内的所有变量

```{r}
## 代码写这里，并运行；
x <- 1
y <- 2
z <- 3
data <- read.table("data/Table1.txt", header = TRUE)
# save.image(file = ".RData")
rm(list = ls())
ls()

load(file = ".RData")
ls()
rm(list = c("x", "y"))
ls()
```

## 练习与作业2：Factor基础

------------------------------------------------------------------------

### **factors 增加**

-   创建一个变量：

`x <- c("single", "married", "married", "single");`

-   为其增加 两个 levels，`single`, `married`;

-   以下操作能成功吗？

`x[3] <- "widowed"`;

-   如果不，请提供解决方案；

```{r}
## 代码写这里，并运行；
x <- c("single", "married", "married", "single")
x <- factor(x, levels = c("single", "married"))
try(x[3] <- "widowed")
# 解决方案
x <- factor(x, levels = c("single", "married", "widowed"))
try(x[3] <- "widowed")
x
```

### **factors 改变**

-   创建一个变量：

`v = c("a", "b", "a", "c", "b")`

-   将其转化为 `factor`，查看变量内容

-   将其第一个levels的值改为任意字符，再次查看变量内容

```{r}
## 代码写这里，并运行；
v <- c("a", "b", "a", "c", "b")
v <- factor(v)
v
levels(v)[1] <- "d"
v
```

-   比较改变前后的 `v` 的内容，改变levels 的操作使 `v` 发生了什么变化？

答：改变levels的操作使v的内容发生了变化，将原来的"a"改为了"d"。

### **factors合并**

-   创建两个由随机大写字母组成的 factors

-   合并两个变量，使其 `factors` 得以在合并后保留

```{r}
## 代码写这里，并运行；
x <- factor(sample(LETTERS, 10, replace = TRUE))
y <- factor(sample(LETTERS, 10, replace = TRUE))
x
y
c(x, y)
```

------------------------------------------------------------------------

### **利用factor排序**

以下变量包含了几个月份，请使用`factor`，使其能按月份，而不是英文字符串排序：

`mon <- c("Mar","Nov","Mar","Aug","Sep","Jun","Nov","Nov","Oct","Jun","May","Sep","Dec","Jul","Nov");`

```{r}
## 代码写这里，并运行；
mon <- c("Mar","Nov","Mar","Aug","Sep","Jun","Nov",
  "Nov","Oct","Jun","May","Sep","Dec","Jul","Nov")
mon <- factor(mon, levels = c("Jan", "Feb", "Mar",
  "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))
sort(mon)
```

------------------------------------------------------------------------

### **forcats的问题**

`forcats`包中的 fct_inorder, fct_infreq和 fct_inseq 函数的作用是什么？

请使用`forcats`包中的`gss_cat`数据举例说明

```{r}
## 代码写这里，并运行；
library(forcats)
head(gss_cat)
# fct_inord er: 按出现顺序为 levels 排序
fct_inorder(gss_cat$marital) %>% levels()
# fct_infreq: 按出现频率为 levels 排序，出现频率高的拍在前面
fct_infreq(gss_cat$marital) %>% levels()
# fct_inseq: 根据 level 的数字大小为 levels 排序，要求 factor levels 为数字
factor(gss_cat$age, levels = 80:60) %>% levels()
factor(gss_cat$age, levels = 80:60) %>% fct_inseq() %>% levels()
```

## 练习与作业3：用 mouse genes 数据做图

------------------------------------------------------------------------

### **画图**

1.  用 readr 包中的函数读取 mouse genes 文件 （从本课程的 Github 页面下载 data/talk04/ ）
2.  选取常染色体（1-19）和性染色体（X，Y）的基因
3.  画以下两个基因长度 boxplot :

-   按染色体序号排列，比如 1, 2, 3 .... X, Y
-   按基因长度 中值 排列，从 短 -\> 长 ...

```{r}
## 代码写这里，并运行；
# 不显示 warning 信息 和 message
options(warn = -1, message = -1)

library(readr)
library(ggplot2)
library(dplyr)
options(warn = 0, message = 0)
mouse_genes <- read_tsv("../data/talk04/mouse_genes_biomart_sep2018.txt",
  col_names = TRUE, show_col_types = FALSE)

colnames(mouse_genes) <- gsub(" ", "_", colnames(mouse_genes))
colnames(mouse_genes) <- gsub("/", "_", colnames(mouse_genes))
colnames(mouse_genes) <- gsub("\\(", "", colnames(mouse_genes))
colnames(mouse_genes) <- gsub("\\)", "", colnames(mouse_genes))
autosome_genes <- filter(mouse_genes, Chromosome_scaffold_name %in% 1:19)

# 按染色体序号排列
ggplot(autosome_genes, aes(x = factor(as.numeric(Chromosome_scaffold_name)),
  y = Transcript_length_including_UTRs_and_CDS)) +
  stat_boxplot(geom = "errorbar", width = 0.3, lwd = 1, color = "lightblue") +
  geom_boxplot(width = 0.5, lwd = 1, color = "lightblue") +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    panel.background = element_blank(),
    panel.border = element_blank(),
    axis.line = element_line(color = "black", size = 0.5),
    axis.ticks = element_line(color = "black", size = 0.5),
    axis.text.x = element_text(size = 10, color = "black"),
    axis.text.y = element_text(size = 10, color = "black"),
    axis.title.x = element_text(size = 12, color = "black"),
    axis.title.y = element_text(size = 12, color = "black"),
    legend.position = "none") +
  coord_cartesian(ylim = c(0, 6000)) +
  stat_summary(fun = mean, geom = "point", shape = 20,
    size = 5, color = "lightgreen", fill = "lightgreen") +
  xlab("Chromosome") +
  ylab("Transcript length (bp)")

# 按基因长度 中值 排列，从 短 -> 长
ggplot(autosome_genes, aes(x = reorder(Chromosome_scaffold_name,
    Transcript_length_including_UTRs_and_CDS, median),
    y = Transcript_length_including_UTRs_and_CDS)) +
  stat_boxplot(geom = "errorbar", width = 0.3, lwd = 1, color = "lightblue") +
  geom_boxplot(width = 0.5, lwd = 1, color = "lightblue") +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    panel.background = element_blank(),
    panel.border = element_blank(),
    axis.line = element_line(color = "black", size = 0.5),
    axis.ticks = element_line(color = "black", size = 0.5),
    axis.text.x = element_text(size = 10, color = "black"),
    axis.text.y = element_text(size = 10, color = "black"),
    axis.title.x = element_text(size = 12, color = "black"),
    axis.title.y = element_text(size = 12, color = "black"),
    legend.position = "none") +
  coord_cartesian(ylim = c(0, 6000)) +
  stat_summary(fun = mean, geom = "point", shape = 20,
    size = 5, color = "lightgreen", fill = "lightgreen") +
  xlab("Chromosome") +
  ylab("Transcript length (bp)")
```